### PR TITLE
Fixes #431

### LINKED ISSUE #431: Project_name column in prep sheet is empty in controls output
Should reflect the project that the blanks/controls refer to.

### PR DESCRIPTION
**Pull request by AmandaBirmingham**
Also provides supporting code that can also be used in future for #483.

@charles-cowart note that the additions made here to `_generate_amplicon_prep_information` should also make it into the new `_generate_metagenomics_prep_information`.

### CODE REVIEW

File: labcontrol/db/process.py [hunk 1]
```diff
@@ -2998,6 +2998,95 @@ def _format_sample_sheet_comments(principal_investigator=None,
...
+            # get unique project names for the part of df defined in the mask
+            curr_unique_projnames = (input_df[plate_non_controls_mask]
+                                  [projname_col_name].unique())
+            if len(curr_unique_projnames) != 1:
+                # Note that we don't error out the first time we find a
+                # plate that doesn't meet expectations; instead we continue to
+                # run through all the plates and identify ALL those that don't
+                # meet expectations.  This way the user can correct all of them
+                # at once.
+                curr_err_msg = "Expected one unique value for plate '{0}' " \
```
Comment (wasade) [inline labcontrol/db/process.py]: There is a heinous edge case here: it is possible that `len(curr_unique_projnames) == 0`. I don't believe it would break the formation of the error message, but the nature of the error is different.

File: labcontrol/db/process.py [hunk 2]
```diff
@@ -2998,6 +2998,95 @@ def _format_sample_sheet_comments(principal_investigator=None,
...
+                problem_plate_messages.append(curr_err_msg)
+            else:
+                # create a mask to define all the rows for this plate where the
+                # projname IS the control value (None); ~ "nots" a whole series
+                plate_controls_mask = plate_mask & (~non_controls_mask)
+
+                # ok to just take first non-control projname because we
+                # verified above there is only one value there anyway
+                input_df.loc[plate_controls_mask, projname_col_name] = \
+                    curr_unique_projnames[0]
```
Comment (wasade) [inline labcontrol/db/process.py]: ....however, if `len(curr_unique_projnames) == 0`, this would throw an `IndexError`. 

If I'm thinking about this correctly, I believe this edge case could result if a plate entirely of controls is run.

File: labcontrol/db/process.py [hunk 3]
```diff
@@ -2998,6 +2998,95 @@ def _format_sample_sheet_comments(principal_investigator=None,
...
+        ------
+        ValueError
+            If any plate contains experimental samples from more (or fewer)
+            than one project.
+        """
+
+        problem_plate_messages = []
+
+        # create a mask to define all the NON-control rows for this plate
+        non_controls_mask = input_df[projname_col_name].notnull()
```
[the remark this answers is not in the document]
Comment (AmandaBirmingham) [reply labcontrol/db/process.py]: Yes, they would implicitly be treated as controls--and would be assigned to the same project as everything else on their plate, which is exactly what the lab wanted (e.g. in the case of the AGP samples that were not in qiita but were plated on an AGP plate and that were *previously* given a project name of "" instead of the AGP project name of the rest of the samples on the plate).

File: labcontrol/db/process.py [hunk 4]
```diff
@@ -2998,6 +2998,95 @@ def _format_sample_sheet_comments(principal_investigator=None,
...
+        for curr_unique_plate in unique_plates:
+            # create a mask to define all the rows for this plate
+            plate_mask = input_df[plate_col_name] == curr_unique_plate
+
+            # create a mask to define all the rows for this plate where the
+            # project name is NOT the control value (None)
+            plate_non_controls_mask = plate_mask & non_controls_mask
+
+            # get unique project names for the part of df defined in the mask
+            curr_unique_projnames = (input_df[plate_non_controls_mask]
```
[the remark this answers is not in the document]
Comment (AmandaBirmingham) [reply labcontrol/db/process.py]: Will do

File: labcontrol/db/process.py (hunk 2 above)
Comment (AmandaBirmingham) [reply labcontrol/db/process.py]: Nope, it won't throw an IndexError in this case--that situation is already taken care of above with the check `len(curr_unique_projnames) != 1` ... if the length is zero, you end up in the error handling and not ever in the `else`.  I have a test to verify this; will check it in in a moment :)

File: labcontrol/db/process.py (hunk 1 above)
Comment (AmandaBirmingham) [reply labcontrol/db/process.py]: That edge case has already been anticipated and is handled by this logic. The error string recorded in that case looks like, e.g., "Expected one unique value for plate 'Test plate 3' but received 0: ".  I agree that the extra colon at the end is a bit ugly but I don't think it is worth refactoring for an edge case that we keep being told will never occur ... :)

File: labcontrol/db/process.py (hunk 2 above)
Comment (wasade) [reply labcontrol/db/process.py]: Oh, wow, sorry. Yes, I misread the flow of logic there.

File: labcontrol/db/process.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wasade) [reply labcontrol/db/process.py]: Thanks

**Comment by AmandaBirmingham** [comment]
Ping @charles-cowart ... do you have an estimate on when you will be available to review this PR?  To sweeten the pot :), I have a fix for #483 ready to go that depends on this PR ... :D

**Comment by charles-cowart** [comment]
@AmandaBirmingham Sorry for the delay, you're ready to go! :)
